File: 0061-rotate-list/0061-rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
 typedef struct ListNode* node;
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (head==NULL||head->next==NULL||k==0){
            return head;
        }
        node end = head;
        int count = 1;
        while (end->next!=NULL){
            end = end->next;
            count++;
        }
        k = k%count;
        k = count - k;
        for (int i = 1; i<=k; i++){
            node temp = head->next;
            end->next = head;
            end = head;
            end->next = NULL;
            head = temp;
        }
        return head;
    }
};

File: 0061-rotate-list/0061-rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if head == None or head.next == None or k == 0:
            return head

        tail = head
        n = 1
        while tail.next != None:
            tail = tail.next
            n += 1

        k = k % n
        if k == 0:
            return head

        tail.next = head
        new_tail_position = n - k - 1
        new_tail = head
        for _ in range(new_tail_position):
            new_tail = new_tail.next

        head = new_tail.next
        new_tail.next = None 
        
        return head
